### PR TITLE
#162949221 Users can follow and unfollow each other

### PR DESCRIPTION
**Pull request by nesh-dev**
#### What does this PR do?

Allow users to follow and unfollow users.
Allow their to view a list of all their followers and followed by

#### Description of Task to be completed?

* Have the following working

- [x] Follow users
- [x] Unfollow users
- [x] List of all followers and unfollows

Use:
 * POST `api/profiles/<username>/follow` 
*  GET `api/articles/<username>/following`
*  GET `api/articles/<username>/followers`

Payload:

#### How should this be manually tested?

1. Clone the repo: `git clone https://github.com/andela/ah-the-unsullied.git`
2. Cd into the cloned repo 
3. Git pull branch ft-user-follow-162949221
4. Set up a virtual environment and install requirements:
      ```
            python3 -m venv venv
            source venv/bin/activate
            pip3 install -r requirements.txt

     ```
5. Set up a .env file in the same level as the venv file and add the following:
      ```
         export SECRET_KEY=""
       export DATABASE_URL=""
      export DJANGO_SETTINGS_MODULE=''
export CLOUDINARY_NAME=''
export CLOUDINARY_KEY=''
export CLOUDINARY_SECRET=''
export HOST_PASSWORD=""
export HOST_USER=""
export SMTP_HOST=''
export EMAIL_PORT='587'
export FACEBOOK_KEY=''
export FACEBOOK_SECRET=''
export TWITTER_KEY=''
export TWITTER_SECRET_KEY=''
export GOOGLE_KEY=''
export GOOGLE_SECRET=''
export ACCESS_TOKEN=''
export SECRET_TOKEN=''
      ```
6. Then run: 
     ``` 
            source .env 
            ./manage.py migrate
            ./manage.py runserver
      ```
7. To test you need to register and login via postman.

* Register endpoint `/api/users`
* Login endpoint `/api/users/login`

#### What are the relevant pivotal tracker stories?
[#162949221](https://www.pivotaltracker.com/n/projects/2235173/stories/162949221)
#### Screenshots (if appropriate)
screenshot 2019-01-30 at 12 48 16
screenshot 2019-01-30 at 12 46 54
screenshot 2019-01-30 at 12 45 56

#### Questions:

### CODE REVIEW

File: authors/tests/test_follow.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        profile = UserProfile.objects.get(user__username='abdi')
+        user_to_check = UserProfile.objects.get(user__username='sam')
+        user_to_check.toggle_follow(profile.user)
+        following_status = user_to_check.check_if_following(profile)
+        self.assertEqual(following_status, False)
+        user_to_check.toggle_follow(profile.user)
+        following_status = user_to_check.check_if_following(profile)
+        self.assertEqual(following_status, True)
+
+
```
Comment (kwanj-k) [inline authors/tests/test_follow.py]: Kindly remove the extra blank lines

**Comment by sylviawanjiku** [comment]
You have an unused function in your code

File: authors/apps/profiles/models.py [hunk 2]
```diff
@@ -9,21 +9,64 @@
...
 
+class UserProfileManager(models.Manager):
+
+    def all(self):
+        queryset = self.get_queryset().all()
+        try:
+            if self.instance:
+                queryset = queryset.exclude(user=self.instance)
+        except:
+            pass
```
Comment (kwanj-k) [inline authors/apps/profiles/models.py]: Hey, just trying to understand why you pass in the case of an exception
Wouldn't you want to catch the error

File: authors/apps/profiles/views.py [hunk 3]
```diff
@@ -1,20 +1,21 @@
+from django.http import JsonResponse
```
Comment (kwanj-k) [inline authors/apps/profiles/views.py]: Unused import

File: authors/apps/profiles/views.py [hunk 4]
```diff
@@ -1,20 +1,21 @@
...
 from rest_framework.generics import ListAPIView
 from rest_framework.permissions import IsAuthenticated
-from rest_framework.response import Response
 from rest_framework import status
 from rest_framework.views import APIView
+from rest_framework.generics import CreateAPIView, DestroyAPIView
```
Comment (kwanj-k) [inline authors/apps/profiles/views.py]: Hey, import the views using line 2 which is importing from generics too

File: authors/tests/test_follow.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+class TestFollow(TestBase):
+
+    def test_follow_yourself(self):
+        """Test handles a case where the user tries to follow themselves"""
+        token = self.authentication_token()
+        url = reverse('profiles:profile_details', kwargs={'username': 'sam'})
+        self.client.credentials(HTTP_AUTHORIZATION='Bearer ' + token)
+        # import pdb;pdb.set_trace()
```
Comment (kwanj-k) [inline authors/tests/test_follow.py]: Remove this coment

File: authors/tests/test_follow.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        self.client.get(url)
+        follow_url = reverse('profiles:follow', kwargs={'username': 'sam'})
+        response = self.client.post(follow_url)
+        # import pdb; pdb.set_trace()
```
Comment (kwanj-k) [inline authors/tests/test_follow.py]: This one too

File: authors/apps/articles/views/articles.py [hunk 7]
```diff
@@ -19,6 +19,7 @@
 
 # local imports
 from authors.apps.articles.models import Article, LikeDislike
+
```
Comment (kwanj-k) [inline authors/apps/articles/views/articles.py]: Do we need this white space?

**Pull request closed without merging**